### DISCUSSION TITLE
Schema Design: One-Many Optional References

### DISCUSSION THREAD

**Discussion by Rykuno** [prisma/prisma#8834] [Q&A] [answered]
I probably butchered this title but hopefully I can explain quickly with a simple example.

I come from using ORMs and Hasura so I'm going to compare approaches to make sure I'm approaching this right in Prisma. I'm reading docs so I'm still learning what Prisma creates under the hood and the most optimal way. 

Would anyone be able to help me understand the `Pro's` and `Con's` to each of the approaches below within Prisma? 

## Scenerio 
A `Guild` and a `User` may each have posts. 

### Method 1

Create a table of `Post` with optional references to both `Guild` and `User`. This is because a `Post` can be either, but not both.

```prisma
model Post {
  id        String      @id @default(uuid())
  text      String
  author    User        @relation("AuthoredPosts", fields: [authorId], references: [id])
  authorId  String
  user      User?       @relation("UserPosts", fields: [userId], references: [id])
  userId    String?
  guild     Guild?      @relation("GuildPosts", fields: [guildId], references: [id])
  guildId   String?
}

model Guild {
  id          String        @id @default(uuid())
  name        String        @unique
  posts       Post[]        @relation("GuildPosts")
}

model User {
  id            String        @id @default(uuid())
  username      String        @unique
  authoredPosts Post[]        @relation("AuthoredPosts")
  posts         Post[]        @relation("UserPosts")
}

```

### Method 2

This is a pattern I'm more familiar with which is to create a reference table.

```prisma
model Post {
  id        String      @id @default(uuid())
  text      String
  author    User        @relation("AuthoredPosts", fields: [authorId], references: [id])
  authorId  String
  guildPosts GuildPost[]
  userPosts  UserPost[]
}

model GuildPost {
  id      String @id @default(uuid())
  post    Post   @relation(fields: [postId], references: [id])
  postId  String
  guild   Guild  @relation(fields: [guildId], references: [id])
  guildId String
}

model UserPost {
  id     String @id @default(uuid())
  post   Post   @relation(fields: [postId], references: [id])
  postId String
  user   User   @relation(fields: [userId], references: [id])
  userId String
}

model User {
  id            String        @id @default(uuid())
  username      String        @unique
  authoredPosts Post[]        @relation("AuthoredPosts")
  UserPost      UserPost[]
}
model Guild {
  id          String        @id @default(uuid())
  name        String        @unique
  GuildPost   GuildPost[]
}
```

**Comment by ryands17** [comment, marked as answer]
@Rykuno 👋 

In the first approach, Prisma automatically creates a reference table so you don't need to create one. As Prisma handles this table internally, you don't need to jump through the reference table to fetch data. You can learn more about implicit and explicit many-to-many relations [here](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
